### PR TITLE
quic: normalize quic idle timeout and handshake timeout

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: Normalize QuicProtocolOptions::idle_timout and  crypto_handshake_timeout to 1ms and 5s respectively if they are configured smaller than those.

Risk Level: low

### CODE REVIEW

File: api/envoy/config/listener/v3/quic_config.proto [hunk 1]
```diff
@@ -31,11 +31,12 @@ message QuicProtocolOptions {
 
   // Maximum number of milliseconds that connection will be alive when there is
   // no network activity. 300000ms if not specified.
-  google.protobuf.Duration idle_timeout = 2;
+  google.protobuf.Duration idle_timeout = 2 [(validate.rules).duration = {gte {nanos: 1000000}}];
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:34]: These are breaking changes of a non-WIP API, which we typically do not allow.
That said, I think these changes make sense. @htuch WDYT?

Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto:34]: QUIC is still in Alpha. Do we allow API change before GA?

Comment (htuch) [reply api/envoy/config/listener/v3/quic_config.proto:34]: If it's in alpha then I suggest marking the protos as WiP. We can't make changes generally to protos such as stronger PGV guards unless there is a `work_in_progress` annotation. I don't think we can do that retrospectively though if folks are already depending on this.

Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto:34]: If so, can we modify their comment and add the enforcement in the code?

Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto]: Modified the code instead. PTAL

File: api/envoy/config/listener/v3/quic_config.proto [hunk 2]
```diff
@@ -30,11 +30,14 @@ message QuicProtocolOptions {
   core.v3.QuicProtocolOptions quic_protocol_options = 1;
 
   // Maximum number of milliseconds that connection will be alive when there is
-  // no network activity. 300000ms if not specified.
+  // no network activity.
+  //
+  // If it is less than 1ms, Envoy will use 1ms. 300000ms if not specified.
```
Comment (adisuissa) [inline api/envoy/config/listener/v3/quic_config.proto:35]: Should the config be rejected in this case?
What I'm worried about is a user that inserts a small value, and is not aware that Envoy uses a different value.

Another question is whether there's a max value?

File: api/envoy/config/listener/v3/quic_config.proto [hunk 3]
```diff
@@ -30,11 +30,14 @@ message QuicProtocolOptions {
...
   google.protobuf.Duration idle_timeout = 2;
 
   // Connection timeout in milliseconds before the crypto handshake is finished.
-  // 20000ms if not specified.
+  //
+  // If it is less than 5000ms, Envoy will use 5000ms. 20000ms if not specified.
```
Comment (ggreenway) [inline api/envoy/config/listener/v3/quic_config.proto:40]: This seems very high for the minimum value. For instance, if the connection is between two hops in a datacenter, a much shorter timeout may be desired; taking away the ability for the user to configure a tight timeout seems unneccessary.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 2 above)
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto:35]: There is no max value. This value shouldn't be too small anyway. Even 1ms is not desirable in real production deployment. If the user configured value is even less than 1ms, the expected behavior oddness would be about the same as 1ms. Would add a warning about the overriding be sufficient?

File: api/envoy/config/listener/v3/quic_config.proto (hunk 3 above)
Comment (danzh2010) [reply api/envoy/config/listener/v3/quic_config.proto:40]: To allow < 5s timeout, we need to expose another interface to allow users to adjust max network idle timeout during handshake in QUICHE, which currently is 5s, to be less than handshake timeout. These values are mostly hardwired for internet facing traffic where QUIC has the most advantage. We don't have any performance data for non-internet facing traffic for QUIC, especially intra-DC. And I'm not sure if QUIC for intra-DC is a proper use case given the low RTT and low packet loss. So IMO we should need the API simple as possible for now and we can always expand if needed.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/config/listener/v3/quic_config.proto]: Yep, code is good, thanks.

File: api/envoy/config/listener/v3/quic_config.proto (hunk 3 above)
Comment (ggreenway) [reply api/envoy/config/listener/v3/quic_config.proto:40]: I see. That makes sense then.

**Comment by danzh2010** [comment]
Ping?

**Review by adisuissa — APPROVED** [review]
/lgtm api
LGTM code changes.

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**